File: solutions/fortran/difference-of-squares/1/difference_of_squares.f90
module difference_of_squares
  implicit none
contains

  ! Square of the sum of first n natural numbers
  integer function square_of_sum(n)
    integer, intent(in) :: n
    integer :: sum_n
    
    if (n <= 0) then
      square_of_sum = 0
      return
    end if
    
    ! Sum of first n natural numbers: n*(n+1)/2
    sum_n = n * (n + 1) / 2
    square_of_sum = sum_n * sum_n
  end function square_of_sum

  ! Sum of squares of first n natural numbers
  integer function sum_of_squares(n)
    integer, intent(in) :: n
    
    if (n <= 0) then
      sum_of_squares = 0
      return
    end if
    
    ! Sum of squares: n*(n+1)*(2n+1)/6
    sum_of_squares = n * (n + 1) * (2*n + 1) / 6
  end function sum_of_squares

  ! Difference between square of sum and sum of squares
  integer function difference(n)
    integer, intent(in) :: n
    
    difference = square_of_sum(n) - sum_of_squares(n)
  end function difference

end module difference_of_squares

File: solutions/python/difference-of-squares/1/difference_of_squares.py
"""Calculate the difference between the square of the sum and the sum of squares."""


def square_of_sum(number: int) -> int:
    """Return the square of the sum of the first `number` natural numbers."""
    sum_of_n = number * (number + 1) // 2
    return sum_of_n ** 2


def sum_of_squares(number: int) -> int:
    """Return the sum of the squares of the first `number` natural numbers."""
    return number * (number + 1) * (2 * number + 1) // 6


def difference_of_squares(number: int) -> int:
    """Return the difference between the square of the sum and the sum of squares."""
    return square_of_sum(number) - sum_of_squares(number)